base/inc/AdePT/SparseVector.h

```diff
@@ -0,0 +1,746 @@
+    if (lnid == group_id) {
+      predmask = votes;
+      cnt      = __popc(votes); // number of votes (set bits) in the subgroup
+      // printf("(warp_id=%d, group_id=%d) cnt = %d  predmask = 0x%x ", warp_id, lnid, cnt, predmask);
+      // print_mask(predmask);
+    }
+  }
+
+// Stage 2: parallel prefix sum of all cnt variables giving the offset for each subgroup
+#pragma unroll
+  for (int i = 1; i < 32; i <<= 1) {
+    int n = __shfl_up_sync(warp_mask, cnt, i);
+    if (lnid >= i) cnt += n;
+  }
+
+  // cnt for lane 31 now contains the sum of all votes in all ballots for warp_id
+
+  // Stage 3: the last lane thread adds the sum of all votes in the warp to get all group offsets.
+  // Each group offset is broadcasted to all threads in the warp using __shfl_sync
+  int group_offset = 0;
+  if (lnid == 31) {
+    // printf("final count for warp %d: cnt = %d\n", warp_id, cnt);
+    group_offset = atomicAdd(nselected, cnt);
+    // printf("warp %d: group_offset = %d  all_votes = %d\n", warp_id, group_offset, *nselected);
+  }
+  group_offset = __shfl_sync(warp_mask, group_offset, 31);
+
+  // Stage 4: Write the indices of the selected elements to the output vector
+  for (int group_id = 0; group_id < group_max; ++group_id) {
+    int mask           = __shfl_sync(warp_mask, predmask, group_id); // broadcast mask stored by lane group_id
+    int subgroup_index = 0;
+    if (group_id > 0)
+      subgroup_index = __shfl_sync(warp_mask, cnt, group_id - 1); // broadcast from thr group_id - 1 if group_id > 0
+
+    if (mask & (1 << lnid)) { // each thr extracts its pred bit
+      int idest     = group_offset + subgroup_index + __popc(mask & ((1 << lnid) - 1));
+      int isrc      = 1024 * warp_id + 32 * group_id + lnid;
+      output[idest] = isrc;
+    }
+  }
+}
+
+/// Dynamic launcher for select_kernel
+template <typename Type, typename Predicate, typename IndexContainer>
+__global__ void select_kernel_launcher(const SparseVectorInterface<Type> *svect, Predicate pred_func,
+                                       IndexContainer *output, unsigned *nselected)
+{
+  const int num_threads = 128;
+  int num_items         = svect->size();
+  int num_blocks        = (num_items + 4095) / 4096; // a warp processes groups of 1024 elements
+  *nselected            = 0;
+  sa_detail::select_kernel<Type, Predicate, IndexContainer>
+      <<<num_blocks, num_threads>>>(svect, pred_func, output, nselected);
+}
+
+/// Make a selection of elements according to the user predicate and write it to the output.
+/// Copy directly elements at the end of the output vector.
+/// Similar to select_kernel, but having extra actions applied to the vectors
+template <typename Type, typename Predicate>
+__global__ void select_and_move_kernel(SparseVectorInterface<Type> *svect, Predicate pred_func,
+                                       SparseVectorInterface<Type> *output, unsigned *nselected, int dest_offset)
```

Comment:
This is very similar to `select_kernel`, any chance we can deduplicate the code?

Reply:
I tried to split the code, unfortunately, one cannot rely on return values since decisions are taken at the thread level, to know what to copy and when. If no immediate solution, we can postpone the refactoring for later.

Reply:
One possibility is to use an extra template parameter, tried that and got some compilation error so decided not to spend extra time on it.

Reply:
Yes, not required in this PR. I might have a look when time allows, but that can happen as a follow-up.